### PR TITLE
Multi-map improvements

### PR DESCRIPTION
**Pull request by Sublimis**
~~* MultiMapDataStore.DataPolicy.RETURN_AUTOFILL: Return data from all datasets in the sequence that support a tile, up to the one that contains data for the entire tile (usually a world map). In other words, once a dataset fills the entire tile, the remaining datasets are not processed for that tile.~~

Here we are solving the following multi-map artifacts:

1. Low-res map coastline peeks out from behind high-res map (RETURN_ALL issue)

2. Incomplete tile returned on map edge (RETURN_FIRST issue; RETURN_ALL has this issue only when zoom levels are limited as in the sample app)

3. Return data from all maps that overlap on a tile

Motivation: Prevent ugly gaps at the edges of the map when the map does not cover the tile completely, and we have a background world map that should easily cover this gap. Additionally, prevent the land areas of the (low-res) world map from "peeking" behind the land areas of a (hi-res) main map, when the main map can provide the entire tile itself.

These problems can be easily seen in the Android application sample MultiMapLowResWorld.

### CODE REVIEW

**Comment by devemux86** [comment]
@Sublimis Can you show screenshots with the difference?

> once a dataset fills the entire tile, the remaining datasets are not processed for that tile

Note that world map supports all tiles BUT does not have all or detailed data.

At least until someone provides a world map with more data:
- https://github.com/mapsforge/mapsforge/issues/319#issuecomment-61953086
- https://github.com/devemux86/cruiser/discussions/101

So the small maps with detailed data are more important in rendering.
World map is only useful at small zoom levels as shown in `MultiMapLowResWorld`.

**Comment by Sublimis** [comment]
> Note that world map supports all tiles BUT does not have all or detailed data.
> 
> At least until someone provides a world map with more data:
> 
>     * [Provide a low detail world map #319 (comment)](https://github.com/mapsforge/mapsforge/issues/319#issuecomment-61953086)
> 
>     * [World map devemux86/cruiser#101](https://github.com/devemux86/cruiser/discussions/101)
> 
> 
> So the small maps with detailed data are more important in rendering. World map is only useful at small zoom levels as shown in `MultiMapLowResWorld`.

Note, when map files are added by priority, world map will be the last to be processed, and only if none of the previous maps provided an entire tile:

https://github.com/mapsforge/mapsforge/blob/f59ea94955f5e0de0fc28c1b9aa9e431c7f850b1/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android/MultiMapLowResWorld.java#L81-L85

**Comment by devemux86** [comment]
It is still not clear what is the difference between before vs after.
For example we see a different map with `master` (not rectangle):

**Comment by devemux86** [comment]
@Sublimis I edited your post with the screenshots to put them together for easy reading.
You check it and see the `img` rule that is useful to avoid filling the posts with large images. :slightly_smiling_face: 
```html
<img src="..." width="...%"/>
```

**Comment by Sublimis** [comment]
> It is still not clear what is the difference between before vs after. For example we see a different map with `master` (not rectangle):
> 

Zoom in a bit more (twice) on the edge of the map, until scale shows "5 km". Our map was taken from here, but the exact map shouldn't matter much: https://www.openandromaps.org/en/downloads/europe

**Comment by devemux86** [comment]
@Sublimis Is is about the rectangle at OAM map edges?
Mapsforge maps do not have it. Where does it come from?

Note: it is easier to test with a small map (with see) like Azores.

**Comment by devemux86** [comment]
@Sublimis Is is about the rectangle at OAM map edges?
Mapsforge maps do not have it.
Where does it come from? Is it in the map data?

Note: it is easier to test with a small map (with sea) like Azores.

**Comment by Sublimis** [comment]
Map source doesn't matter, as long as you find zoom level >10 and an edge with incomplete tile. Ok, we can try Azores or anything else.

**Comment by devemux86** [comment]
@Sublimis Try to also reproduce it with Mapsforge maps.

**Comment by Sublimis** [comment]
No problem 🙂  Azores map from mapsforge server:

| 20241126-azores-before1 | 20241126-azores-before2 |
| --------------------------------------- | --------------------------------------- |

**Comment by Sublimis** [comment]
> No problem 🙂 Azores map from mapsforge server:

The second image just shows the zoom center!

**Comment by Sublimis** [comment]
With this PR applied:


| 20241126-azores-after3 |  |
| --------------------------------------- | --------------------------------------- |

**Comment by devemux86** [comment]
@Sublimis Thanks, I understand now, I was looking at different things! :slightly_smiling_face:

**Comment by devemux86** [comment]
@Sublimis 

- Does the order of opened map files matter?
As we shouldn't expect apps to open multiple map files in a specific order.

- `RETURN_AUTOFILL` can be bit confusing as name (especially "auto").
Is there a better name?

- Can you also add this flag in desktop Samples?

**Comment by devemux86** [comment]
@Sublimis What happens with adjacent maps where each map has a part of the data of their shared tile?
Does `break` only return the data of one map?

**Comment by devemux86** [comment]
@Sublimis 

> Does the order of opened map files matter?
As we shouldn't expect apps to open multiple map files in a specific order.

The order of opened maps matters, if we add the world map first, then it covers everything.

> What happens with adjacent maps where each map has a part of the data of their shared tile?
Does `break` only return the data of one map?

Indeed, this does not seem to work properly.

For example, opening the maps of Austria and Switzerland in this order.
With `AUTOFILL` their shared tile is filled with only Austria data compared to `ALL`:

|AUTOFILL|
|-|
|AUTOFILL|

|ALL|
|-|
|ALL|

**Comment by Sublimis** [comment]
>     * Does the order of opened map files matter?

Yes, maps should be inserted in order of precedence for this to function as advertised. Or at the very least, the world map (if used) should be inserted last. Anyone using the RETURN_FIRST is already doing this. If a user who *consciously chooses to use the RETURN_AUTOFILL flag* sees artifacts like the above, they will simply change the order of the inserted maps. (This should be put as a note in the description of the flag, if not already clear enough).

>       As we shouldn't expect apps to open multiple map files in a specific order.

This is incorrect. Anyone who uses the RETURN_FIRST flag already pays attention to the order of the maps inserted if there is a chance that more than one map can serve a particular tile, which is often the case. Especially if they are using the world map, which overlaps all the other maps.

>     * `RETURN_AUTOFILL` can be bit confusing as name (especially "auto").
>       Is there a better name?

Since the library decides ("automatically") when a tile is "filled", the name seems appropriate. We don't have a better name, if anyone can think of a better name, please suggest it!

>     * Can you also add this flag in desktop Samples?

Yes

**Comment by Sublimis** [comment]
> The order of opened maps matters, if we add the world map first, then it covers everything.
> 
> > What happens with adjacent maps where each map has a part of the data of their shared tile?
> > Does `break` only return the data of one map?
> 
> Indeed, this does not seem to work properly.

It is working exactly as advertised. Note, this flag is not a replacement for RETURN_ALL!

**Comment by devemux86** [comment]
@Sublimis Indeed `RETURN_FIRST` is not really very useful, as it produces half filled multi maps.
Why open the maps of Austria and Switzerland and see only Austria on their borders?

Also this new flag seems to have a very specific use, like world map + 1 country on top?
As seen above multiple countries will not work properly.

Please let's revert the samples to `master` code, as `ALL` is what most users would use.

**Comment by Sublimis** [comment]
For your example above to work "correctly" with the AUTOFILL flag (i.e. to return data from both maps), neither map should claim to have data for an entire tile that it doesn't actually have. But that then becomes "just" a question of implementation details, i.e. how to get the exact map boundaries (currently this is a rough rectangle). That's why this flag is currently most useful with a world map.

**Comment by devemux86** [comment]
@Sublimis We also need to think about the performance of the map.
Heavy calculations with thousands of map data on shared tiles can have negative results.

**Comment by Sublimis** [comment]
> @Sublimis Indeed `RETURN_FIRST` is not really very useful, as it produces half filled multi maps. Why open the maps of Austria and Switzerland and see only Austria on their borders?

It seems you should think this through some more. Do take your time. RETURN_FIRST is useful when the user knows the maps they throw into the `MultiMapDataStore` are disjoint, i.e. not overlapping anywhere (if not using the world map). If they use the world map, this will also work, as the user will see their country as expected, but when they go to entirely different part of the world they will see the world map.

**Comment by Sublimis** [comment]
Also note that no one is forcing anyone to use the new flag. It has its own benefit that no other flag covers. They all have artifacts in a common useful scenario as shown above.

The new flag is in addition to the current flags and does not change their current behavior in any way.

The sample application can be reverted to RETURN_ALL, no problem. However, there is no performance difference between the two flags for the sample application.

**Comment by Sublimis** [comment]
> @Sublimis We also need to think about the performance of the map. Heavy calculations with thousands of map data on shared tiles can have negative results.

This doesn't really have anything to do with the new flag, does it? It may only be faster than RETURN_ALL, or exactly the same. The new flag can never be slower than RETURN_ALL.

**Comment by devemux86** [comment]
@Sublimis I meant that users usually open maps of neighboring countries.
Of course there could be other use cases of the map library. :slightly_smiling_face:

**Comment by Sublimis** [comment]
By the way, we first tried to implement such functionality at the "layer level", i.e. use RETURN_ALL (without limiting the zoom level of the world map), but somehow force the sea layer of the more important map to be drawn on top of the less important (world) map, to prevent coastline artifacts. However, we learned that ALL sea layers are hard-coded as layer zero, so no go. Maybe this is a part of the library that could also be improved in the future, but we will almost certainly not deal with it.

We also first used the name RETURN_SEQUENTIAL for the new flag, but this name doesn't really say much. That's why there are some `boolean sequential` parameters in the code, in case you were wondering. We should probably rename that.

The RETURN_AUTOFILL flag is perhaps best described as a "fixed" version of RETURN_FIRST; it behaves the same unless the tile is incomplete (partially drawn), in which case it will try to load more data (unlike RETURN_FIRST).

Also, our opinion is that no one should ever look at a map that doesn't have a world background when zoomed out, especially considering that the current world map is so lightweight. But that's just us 🙂

**Comment by Sublimis** [comment]
Let's recap and propose a slightly different, more universal approach.

Here we are solving the following multi-map artifacts:

1. Low-res map coastline peeks out from behind high-res map (RETURN_ALL issue)

2. Incomplete tile returned on map edge (RETURN_FIRST issue; RETURN_ALL has this issue only when zoom levels are limited as in the sample app)

Now we also want to satisfy this useful condition:

3. Return data from all maps that overlap on a tile

The proposal is to provide the ability to assign priority to the inserted map. By default, this can be 0 for all maps via method overloading (so that no code changes are required from the user). Then return all data (just like RETURN_ALL would do), but sorted by priority. If a map has priority <0, return its data only if the tile is not yet filled by other maps. We can call this a background map priority.

Such functionality can even be implemented without introducing a new flag. By default it will behave just like the RETURN_ALL is behaving right now.

However, we could also introduce a small auxiliary functionality so that users of the RETURN_ALL + world map can immediately benefit from removing the "peeking coastline" artifact: The library could detect when a map is called "world.map" and automatically assign it a negative priority.

What do you think?

**Comment by devemux86** [comment]
@Sublimis Thanks for the detailed description.

A universal solution seems better, so opening **azores** + **austria** + **switzerland** + **world**
would work correctly in all cases:
- Coastline of Azores
- Common tiles of Austria and Switzerland

**Comment by devemux86** [comment]
@Sublimis It is also important if we manage not to break how the public API works.
As it is already used in many applications.

**Comment by devemux86** [comment]
@Sublimis Can you change the title and the description of the pull request?

**Comment by devemux86** [comment]
@Sublimis It does not work if we don't set priority:
```java
Caused by: java.lang.IllegalArgumentException: Duplicate map database
                 	at org.mapsforge.map.datastore.MultiMapDataStore.addMapDataStore(MultiMapDataStore.java:70)
                 	at org.mapsforge.samples.android.MultiMapLowResWorld.onCreate(MultiMapLowResWorld.java:86)
```

File: mapsforge-map/src/main/java/org/mapsforge/map/datastore/MultiMapDataStore.java [hunk 1]
```diff
@@ -44,13 +48,13 @@ public enum DataPolicy {
...
     private final DataPolicy dataPolicy;
-    private final List<MapDataStore> mapDatabases;
+    private final Set<MapDataStore> mapDatabases;
     private LatLong startPosition;
     private byte startZoomLevel;
 
     public MultiMapDataStore(DataPolicy dataPolicy) {
         this.dataPolicy = dataPolicy;
-        this.mapDatabases = new ArrayList<>();
+        this.mapDatabases = new TreeSet<>();
```
Comment (devemux86) [inline mapsforge-map/src/main/java/org/mapsforge/map/datastore/MultiMapDataStore.java:57]: @Sublimis Why this change?

If the order of insertion is still important, we can sort the list in the app or in the library.

**Comment by Sublimis** [comment]
There, that should do it.


image

File: mapsforge-map/src/main/java/org/mapsforge/map/datastore/MultiMapDataStore.java (hunk 1 above)
Comment (Sublimis) [reply mapsforge-map/src/main/java/org/mapsforge/map/datastore/MultiMapDataStore.java]: Done that. Collections.sort() is guaranteed to be stable, so it will not change the insertion order if all map files happen to have the same priority (which is the default, so it retains the current behavior).

**Comment by Sublimis** [comment]
> @Sublimis It does not work if we don't set priority. Also desktop samples does not work.
> 
> ```java
> java.lang.IllegalArgumentException: Duplicate map database
>     at org.mapsforge.map.datastore.MultiMapDataStore.addMapDataStore(MultiMapDataStore.java:70)
>     at org.mapsforge.samples.android.MultiMapLowResWorld.onCreate(MultiMapLowResWorld.java:86)
> ```

Fixed that, please try now with the latest code.

File: mapsforge-map-reader/src/main/java/org/mapsforge/map/reader/MapFile.java [hunk 2]
```diff
@@ -237,6 +238,12 @@ public MapFile(File mapFile, String language) {
             this.databaseIndexCache = new IndexCache(this.inputChannel, INDEX_CACHE_SIZE);
 
             this.timestamp = mapFile.lastModified();
+
+            // 2024: This may be removed in the future when world.map users get used to the map priorities feature
+            if ("world.map".equalsIgnoreCase(mapFile.getName())) {
+                // Some random (very) negative value no one will use
+                setPriority(Integer.MIN_VALUE + 374);
+            }
```
Comment (devemux86) [inline mapsforge-map-reader/src/main/java/org/mapsforge/map/reader/MapFile.java:246]: @Sublimis We should not do this inside the library.
This must be done by the apps that know the map content.

Also, the map names are not fixed, they can be changed freely.
For example [paws.cz](https://osm.paws.cz/) has europe_admin map.

Comment (devemux86) [reply mapsforge-map-reader/src/main/java/org/mapsforge/map/reader/MapFile.java:246]: @Sublimis Let's put it only in the sample apps.
Android has it, we can also put it in desktop samples.

Comment (devemux86) [reply mapsforge-map-reader/src/main/java/org/mapsforge/map/reader/MapFile.java:246]: @Sublimis We can set it manually on desktop samples as well, just like on Android.
They are just examples.

Comment (Sublimis) [reply mapsforge-map-reader/src/main/java/org/mapsforge/map/reader/MapFile.java]: It's actually better like this, without assuming that the world.map file should be present, so the user decides which files to use. Because desktop sample has no startup menu like the Android app, it's just a single app.

**Comment by devemux86** [comment]
@Sublimis Now it seems to work ok in all cases.
If it is complete, we can merge it?

**Comment by Sublimis** [comment]
Wait, maybe to clarify comments a bit.

**Comment by Sublimis** [comment]
Done.
